File: cracking-code-interview/src/main/java/leetcode/dp/LongestCommonSubsequence.java
package leetcode.dp;

public class LongestCommonSubsequence {
  public int longestCommonSubsequence(String text1, String text2) {
    if (text1 == null || text2 == null || text1.isEmpty() || text2.isEmpty()) return 0;

    var memo = new int[1001][1001];
    for (var i = 0; i < text1.length(); ++i) {
      for (var j = 0; j < text2.length(); ++j) {
        memo[i + 1][j + 1] =
            text1.charAt(i) == text2.charAt(j)
                ? memo[i][j] + 1
                : Math.max(memo[i + 1][j], memo[i][j + 1]);
      }
    }

    return memo[text1.length()][text2.length()];
  }

  public static void main(String[] args) {
    // "bsbininm"
    // "jmjkbkjkv"

    // "abcba"
    // "abcbcba"
    System.out.println(
        new LongestCommonSubsequence().longestCommonSubsequence("bsbininm", "jmjkbkjkv"));
    System.out.println(new LongestCommonSubsequence().longestCommonSubsequence("abcba", "abcbcba"));
    System.out.println(new LongestCommonSubsequence().longestCommonSubsequence("aaaaa", "aaaa"));
  }
}

File: leet-kotlin/src/main/kotlin/algo/dp/LongestCommonSubsequence.kt
package algo.dp

import kotlin.math.max

fun longestCommonSubsequence(text1: String, text2: String): Int {
    if (text1.isEmpty() || text2.isEmpty()) return 0

    val memo = Array(text1.length + 1) { IntArray(text2.length + 1) }
    for (i in text1.indices) {
        for (j in text2.indices) {
            memo[i + 1][j + 1] =
                if (text1[i] == text2[j]) memo[i][j] + 1
                else max(memo[i + 1][j], memo[i][j + 1])
        }
    }

    return memo[text1.length][text2.length]
}


fun main() {
    println(longestCommonSubsequence("ezupkr", "ubmrapg"))
}